Fix #84 - Revise development documentation (#158)
nautobot/nautobot: nautobot/docs/development/getting-started.md

# Getting Started

## Git Branches

> The Nautobot project utilizes three persistent git branches to track work:
> 
> * `main` - Serves as a snapshot of the current stable release
> * `develop` - All development on the upcoming stable release occurs here
> * `feature` - Tracks work on an upcoming major release
> 
> You will always base pull requests off of the `develop` branch, or off of `feature` if you're working on a new major release. **Never** merge pull requests into the `main` branch, which receives merges only from the `develop` branch.

Reviewer: ```suggestion
The Nautobot project follows a branching model based on [Git-flow](https://nvie.com/posts/a-successful-git-branching-model/). As such, there are two persistent git branches:

* `main` - Serves as a snapshot of the current stable release
* `develop` - All development on the upcoming stable release occurs here

At any given time, there may additionally be zero or more long-lived branches of the form `develop-X.Y.Z`, where `X.Y.Z` is a future stable release later than the one currently being worked on in the main `develop` branch.

You will always base pull requests off of the `develop` branch, or off of `develop-X.Y.Z` if you're working on a feature targeted for a later release. **Never** target pull requests into the `main` branch, which receives merges only from the `develop` branch.
```

NOTICE           dist         nautobot    scripts
```

### About Remote Repos

Git refers to remote repositories as *remotes*. When you make your initial clone of your fork, Git defaults to naming this remote `origin`. Throughout this documentation, the following remote names will be used:

- `origin` - The default remote name used to refer to *your fork of Nautobot* 
- `upstream` - The main remote used to refer to the *official Nautobot repository*

> ### Setting up your Remotes

Reviewer: It's not clear to me actually what we gain by adding this section. Are you planning to add more content about pulling from upstream as needed, etc.? As written currently I'm not seeing where we're using the `upstream` remote at all.

Author: Yeah. That's what I mentioned in the git commit. :P

Reviewer: :reading-is-fundamental:

Author: I'll add a short section on branching to reference it.

origin	git@github.com:yourusername/nautobot.git (push)
upstream	git@github.com:nautobot/nautobot.git (fetch)
upstream	git@github.com:nautobot/nautobot.git (push)
```

You're now ready to proceed to the next steps.

!!! hint
	You will always **push** changes to `origin` (your fork) and **pull** changes from `upstream` (official repo).

### Creating a Branch

Before you make any changes, always create a new branch. In the majority of cases, you'll always want to create your branches from the `develop` branch.

Before you ever create a new branch, always  checkout the `develop` branch and make sure you you've got the latest changes from `upstream`.

```no-highlight
$ git checkout develop
$ git pull upstream develop
```

!!! warning
	If you do not do this, you run the risk of having merge conflicts in your branch, and that's never fun to deal with. Trust us on this one.

Now that you've got the latest upstream changes, create your branch. It's convention to always prefix your branch name with your GitHub username, separated by hyphens. For example:

```no-highlight
$ git checkout -b yourusername-myfeature
```

## Enabling Pre-Commit Hooks

> Nautobot ships with a [Git pre-commit hook](https://githooks.com/) script that automatically checks for style compliance and missing database migrations prior to committing changes. This helps avoid erroneous commits that result in CI test failures. You are encouraged to enable it by creating a link to `scripts/git-hooks/pre-commit`:

Reviewer: The pre-commit hook pretty much assumes you're using the virtualenv workflow. We should look into what would be required to make it work with the docker-compose workflow as well, but in the meantime, it might be good to add a disclaimer here.

Author: Ah yes, it would be good but if it's assuming those are installed locally, there would need to be another check or setting in the pre-commit-hook script to assert that can be ran from within the container as well.

Author: I found a way to get it to detect whether it's in side a virtualenv but would rather tackle this in another PR. Putting the disclaimer in there for now.

```no-highlight
$ cd .git/hooks/
$ ln -s ../../scripts/git-hooks/pre-commit
```

## Setting up your Development Environment

Getting started with Nautobot development is pretty straightforward, and should feel very familiar to anyone with Django development experience. We can recommend either a [Docker Compose workflow](#docker-compose-workflow) (if you don't want to install dependencies such as PostgreSQL and Redis directly onto your system) or a [Python virtual environment workflow](#python-virtual-environment-workflow).

### Docker Compose Workflow

This workflow uses [Docker](https://docs.docker.com/get-docker/) and [Docker Compose](https://docs.docker.com/compose/) and assumes that you have them installed.

For the Docker Compose workflow, Nautobot uses [Invoke](http://docs.pyinvoke.org/en/latest/index.html) as a replacement for Make. Invoke was chosen because it is less arcane than make. Instead of a `Makefile`, Invoke reads the `tasks.py` in the project root.

#### Install Invoke

> Because it is used to execute all common Docker workflow tasks, Invoke must be installed globally for your user environment.

Reviewer: We shouldn't be recommending to globally install any Python packages, not even Invoke - we should either:
- direct the user to do a user install (`pip install --user invoke`) and ensure that their `$PATH` is set correctly
- or direct the user to create a virtualenv and install Invoke there,
- or install Poetry and use Poetry to create a virtualenv containing Invoke.

Author: `pip install --user` is the default these days unless you explicitly install as root or a user that has permissions to modify the system-installed version.

Really the goal is "Invoke must be installed for use outside of virtualenvs" similar to that of Poetry. And then we should describe them in the same way.

Reviewer: OK. I was thrown by the "installed globally" wording.

Author: Yeah, I'll flesh it out.

  build             Build all docker images.
  cli               Launch a bash shell inside the running Nautobot container.
  coverage-report   Run coverage report
  coverage-run      Run tests
  createsuperuser   Create a new superuser in django (default: admin), will prompt for password.
  debug             Start Nautobot and its dependencies in debug mode.
  destroy           Destroy all containers and volumes.
  flake8            Check PEP8 compliance and other style issues
  makemigrations    Run Make Migration in Django.
  migrate           Perform migrate operation in Django.
  nbshell           Launch a nbshell session.
  start             Start Nautobot and its dependencies in detached mode.
  stop              Stop Nautobot and its dependencies.
  tests             Run all tests
  vscode            Launch Visual Studio Code with the appropriate Environment variables to run in a container.
```

#### Using Docker with Invoke

A development environment can be easily started up from the root of the project using the following commands:

- `invoke build` - Builds Nautobot docker images
- `invoke createsuperuser` - Creates a superuser account for the Nautobot application

> - `invoke debug` - Starts Linux system, PostgreSQL, Redis, and Nautobot in debug mode and attaches their output to the terminal in the foreground. You may enter Control-C to stop the containers.

Reviewer: ```suggestion
- `invoke debug` - Starts Docker containers for Nautobot, PostgreSQL, Redis, and the Jobs worker in debug mode and attaches their output to the terminal in the foreground. You may enter Control-C to stop the containers.
```

Additional useful commands for the development environment:

- `invoke start` - Starts all Docker containers to run in the background with debug disabled
- `invoke stop` - Stops all containers created by `invoke start`

#### Working with Docker Compose

> The files related to the Docker container environment used for development can be found inside of the `development` directory a the root of the project.

Reviewer: ```suggestion
The files related to the Docker container environment used for development can be found inside of the `development` directory at the root of the project.
```

```

!!! warning
    Please name the **.env** file ``override.env`` to prevent credentials from accidentally being checked into Git as ``override.env`` is set in the ``.gitignore`` file.

> These variables will tell `docker-entrypoint.sh` to create the superuser with the specified username, email, password, and API token.

Reviewer 2: The `override.env` variables will signal to that `docker-entrypoint.sh`shell script to create ....

##### Special Workflow for Containers on Remote Servers

A slightly different workflow is needed when your development container is running on a remotely-connected server (such as with SSH). VS Code, will not offer the **Reopen in Container** option on a remote server.

To work with remote containers, after `invoke build` use `docker-compose` as follows to start the containers. This prevents the HTTP service from automatically starting inside the container:

```no-highlight
$ cd development
$ docker-compose -f docker-compose.yml -f docker-compose.debug.yml up
```

- Now open the VS Code Docker extension. In the `CONTAINERS/development` section, right click on a running container and select the **Attach Visual Studio Code** menu item.
- The **Select the container to attach VS Code** input field provides a list of running containers.
- Click on `development_nautobot_1` to use VS Code inside the container. The `devcontainer` will startup now. 
- As a last step open the folder `/opt/nautobot` in VS Code.

### Python Virtual Environment Workflow

> This workflows uses Python and Poetry to work with your development environment locally. It requires that you install the requireed system dependencies on your system yourself.

Reviewer: ```suggestion
This workflow uses Python and Poetry to work with your development environment locally. It requires that you install the required system dependencies on your system yourself.
```

#### Microsoft Visual Studio Code Integration

For users of Microsoft Visual Studio Code, several files are included to ease development and integrate with the [VS Code Remote - Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers). The following related files are found relative to the root of the project:

- `.devcontainers/devcontainer.json` - Dev. container definition
- `nautobot.code-workspace` - VS Code workspace configuration for Nautobot
- `development/docker-compose.debug.yml` - Docker Compose file with debug configuration for VS Code

After opening the project directory in VS Code in a supported environment, you will be prompted by VS Code to **Reopen in Container** and **Open Workspace**. Select **Reopen in Container** to build and start the development containers. Once your window is connected to the container, you can open the workspace which enables support for Run/Debug.

To start Nautobot, select **Run Without Debugging** or **Start Debugging** from the Run menu. Once Nautobot has started, you will be prompted to open a browser to connect to Nautobot.

##### Special Workflow for Containers on Remote Servers

> A slightly different workflow is needed when your development container is running on a remotely-connected server (such as with SSH). VS Code, will not offer the **Reopen in Container** option on a remote server.

Reviewer: ```suggestion
A slightly different workflow is needed when your development container is running on a remotely-connected server (such as with SSH). VS Code will not offer the **Reopen in Container** option on a remote server.
```

It may not always be convenient to enter into the virtual shell just to run programs. You may also execute a given command ad hoc within the project's virtual shell by using `poetry run`:

```

> $ poetry run mkdocs server

Reviewer: ```suggestion
$ poetry run mkdocs serve
```

Author: 🤦

It may not always be convenient to enter into the virtual shell just to run programs. You may also execute a given command ad hoc within the project's virtual shell by using `poetry run`:

```
$ poetry run mkdocs serve
```

Check out the [Poetry usage guide](https://python-poetry.org/docs/basic-usage/) for more tips.	

!!! note

> 	Unless otherwise noted, all following commands will assumed to be ran inside the virtualenv. 

Reviewer: ```suggestion
	Unless otherwise noted, all following commands should be executed inside the virtualenv. 
```

$ nautobot-server test
```

In cases where you haven't made any changes to the database (which is most of the time), you can append the `--keepdb` argument to this command to reuse the test database between runs. This cuts down on the time it takes to run the test suite since the database doesn't have to be rebuilt each time. 

!!! note

> 	Note that this argument will cause errors if you've modified any model fields since the previous test run.

Reviewer: ```suggestion
	Using the `--keepdb` argument will cause errors if you've modified any model fields since the previous test run.
```
